### ISSUE TITLE
Idea: Trying out Atomic Style Output

### ISSUE THREAD

**Issue by rtfeldman** [rtfeldman/elm-css#341]
It's possible that Atomic CSS-style output (each declaration gets its own classname) would be better than doing hashed classes. It's not clear if this is a good idea, but it seems worth considering.

Design gotchas to watch out for: (some helpful replies [to this tweet](https://twitter.com/rtfeldman/status/934304637137670144))

1. What do we do in the case of custom properties generated with `property`? What do they compile to? How do we make sure that what they compile to does not suffer from https://github.com/rtfeldman/elm-css/issues/337?

2. Consider `#2` in http://fela.js.org/docs/introduction/Caveats.html - probably the solution is to (1) make `border3` be a "mixin" that specifies `borderWidth`, `borderColor`, etc, (2) pre-resolve cascades, e.g. if I say `css [ borderWidth (px 2), borderWidth (px 3), borderWidth (px 4) ]` I get only one property in the output: `border-width: 4px`. Maybe we should do this anyway?

3. In the "compile to a hash" world, people can't rely on CSS classes at all for e2e testing. Great! However, in this world, classnames are predictable enough that maybe people would.

4. Attempting to memoize classname calculations using a `Dict` would almost certainly be a net loss from a perf perspective. `Dict` lookups take logarithmic time, whereas string concatenation is constant. So we shouldn't expect the same level of perf that Styletron's benchmarks show.

How to benchmark:

1. Take a large page that's using the existing output, then recompile using the version with atomic style output. Compare load times for both. At the end of the day, this matters more than microbenchmarks - all things considered, did it make the page faster or slower?

2. Try it with SSR as well. Wouldn't this increase downloaded HTML file size? Maybe it's fine with gzip though, if critical CSS is in <style> in the <head>?

**Comment by Stijnkool** [comment]
- When each declaration gets its own class, how would cascading work when there are multiple declarations for the same css poperty? Should all those declarations be collapsed into the one the gets priority? In that case, also collapse custom properties with the same name into one, and there should not be an issue concerning https://github.com/rtfeldman/elm-css/issues/337.
- How do media queries cascade when generating 1 class per declaration? How is the order of the output determined? Styletron also seems to suffer from this. (https://github.com/rtsao/styletron/issues/113) Their solution is to not use overlapping mediaQueries and just use disjoint ones. It's a fix that is not inline with how I think people usually write media queries. But it is a fix nonetheless. 
- I think there are 2 options concerning this particular topic:
  - Don't care about the order of the output and fix it in userland.
  - Care about the order of the output and guarantee in some way that when defining overlapping media queries on the same element, they will be outputted in the order they are defined.

**Comment by rtfeldman** [comment]
> When each declaration gets its own class, how would cascading work when there are multiple declarations for the same css poperty? Should all those declarations be collapsed into the one the gets priority?

The last one should win, I suppose unless an earlier one uses `!important`. So if I write `css [ color blue, color green ]` it should only emit the equivalent of `color green`, but if I write  `css [ color blue |> important, color green ]`, then it should only emit the equivalent of `color blue`.

https://github.com/rtfeldman/elm-css/pull/345 fixed #337 by changing to a "last one wins" policy regarding multiple usages of `css` inside a single attribute list, so that case can't cause problems anymore. 😅

**Comment by rtfeldman** [comment]
> How do media queries cascade when generating 1 class per declaration? How is the order of the output determined? Styletron also seems to suffer from this. (rtsao/styletron#113) Their solution is to not use overlapping mediaQueries and just use disjoint ones. It's a fix that is not inline with how I think people usually write media queries. But it is a fix nonetheless.

Yeah, that seems pretty bad. I'm not sure if it's a deal-breaker, but it is very close to one in my mind. 🤔
